src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+#define _GNU_SOURCE 1
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <stdlib.h>
+#include <sys/un.h>
+#include <errno.h>
+#include <err.h>
+
+#define N_ELEMENTS(arr) (sizeof (arr) / sizeof ((arr)[0]))
+#define SD_LISTEN_FDS_START 3
+#define MAX_COCKPIT_WS_ARGS 5
+
+static struct instance_type
+{
+  const char *sockname;
+  const char *argv[MAX_COCKPIT_WS_ARGS];
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
+
+static void
+handle_cockpit_ws_exited (int sigid)
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+              return;
+            }
+        }
+    }
+  /* This can only be reached in case cockpit-ws exited fast and the parent
+   * process did not manage to insert its PID into the array
+   */
+  errx (1, "Could not find the process in socket_to_pid array");
+}
+
+static int
+spawn_cockpit_ws (const char *ws_path, int fd,
+                  struct instance_type sock_instance)
+{
+  pid_t pid = fork ();
+
+  if (pid == -1)
+    {
+      err (1, "fork() failed");
+    }
+  else if (pid > 0)
+    {
+      return pid;
+    }
+  else
+    {
+      const char *args[MAX_COCKPIT_WS_ARGS + 1];
+      char *env[] = { NULL };
+      int i = 0;
+
+      /* pass the socket to ws like systemd activation does, see sd_listen_fds(3) */
+      if (dup2 (fd, SD_LISTEN_FDS_START) < 0)
+        err (1, "dup2() failed");
+
+      for (i = 0; sock_instance.argv[i] != NULL; i++)
+        args[i + 1] = sock_instance.argv[i];
+      args[0] = ws_path;
+      args[i + 1] = NULL;
+
+      if (execve (ws_path, (char **) args, env) < 0)
+        errx (1, "spawning cockpit-ws instance failed");
+
+      exit (0);
```

Comment:
I guess this will actually never be reached.

The check for the error on `execve()` above is strange: if `execve()` is successful then it will never return, so the return value will always indicate failure (in which case you'll errx, accordingly).  As such, I'd probably get rid of the `if` and the `exit` and just unconditionally `err()` here.